be/src/vec/columns/column_dictionary.h

```diff
@@ -0,0 +1,381 @@
+        end_ptr += data_num * sizeof(T);
+        codes.set_end_ptr(end_ptr);
+    }
+
+    bool is_dict_inited() const { return dict_inited; }
+
+    bool is_dict_sorted() const { return dict_sorted; }
+
+    bool is_dict_code_converted() const { return dict_code_converted; }
+
+    ColumnPtr convert_to_predicate_column() {
+        auto res = vectorized::PredicateColumnType<StringValue>::create();
+        size_t size = codes.size();
+        res->reserve(size);
+        for (size_t i = 0; i < size; ++i) {
+            auto& code = reinterpret_cast<T&>(codes[i]);
+            auto value = dict.get_value(code);
+            res->insert_data(value.ptr, value.len);
+        }
+        dict.clear();
+        return res;
+    }
+
+    void convert_dict_codes() {
+        if (!is_dict_sorted()) {
+            sort_dict();
+        }
+
+        if (!is_dict_code_converted()) {
+            for (size_t i = 0; i < size(); ++i) {
+                codes[i] = dict.convert_code(codes[i]);
+            }
+            dict_code_converted = true;
+        }
+    }
+
+    void sort_dict() {
+        dict.sort();
+        dict_sorted = true;
+    }
+
+    class Dictionary {
+    public:
+        Dictionary() = default;
+
+        void reserve(size_t n) {
+            dict_data.reserve(n);
+            inverted_index.reserve(n);
+        }
+
+        inline void insert_value(StringValue& value) {
+            dict_data.push_back_without_reserve(value);
+            inverted_index[value] = inverted_index.size();
+        }
+
+        inline T find_code(const StringValue& value) const {
+            auto it = inverted_index.find(value);
+            if (it != inverted_index.end()) {
+                return it->second;
+            }
+            return -1;
+        }
+
+        inline T find_bound_code(const StringValue& value, bool lower, bool eq) const {
+            if (lower) {
+                return std::lower_bound(dict_data.begin(), dict_data.end(), value) - dict_data.begin() - eq;
+            } else {
+                return std::upper_bound(dict_data.begin(), dict_data.end(), value) - dict_data.begin() + eq;
+            }
+        }
+
+        inline phmap::flat_hash_set<T> find_codes(const phmap::flat_hash_set<StringValue>& values) const {
+            phmap::flat_hash_set<T> code_set;
+            for (const auto& value : values) {
+                auto it = inverted_index.find(value);
+                if (it != inverted_index.end()) {
+                    code_set.insert(it->second);
+                }
+            }
+            return code_set;
+        }
+
+        inline StringValue& get_value(T code) { return dict_data[code]; }
+
+        void clear() {
+            dict_data.clear();
+            inverted_index.clear();
+            code_convert_map.clear();
+        }
+
+        void sort() {
+            size_t dict_size = dict_data.size();
+            std::sort(dict_data.begin(), dict_data.end(), comparator);
```

Comment:
How cost is sorting here?

Reply:
There is no specific test for sort. However, according to the SSB single-table Q2.2 query (the dictionary size is 1000), after sorting and storing codes, it is faster than the original method.

The original way:
- ReadColByIdxTime: 2s957ms
- ShortPredEvalTime: 5s10ms

After low cardinality optimization:
- ReadColByIdxTime: 2s556ms      (Dictionary sorting at this stage)
- ShortPredEvalTime: 481.672ms